messages/compiler/python/test_pygen.py

```diff
@@ -0,0 +1,181 @@
+import pytest
+import sys
+
+sys.path.append("..")
+import cmfc
+from python import pygen
+from python.serialize import CmfSerializeError
+
+
+def gen_ast():
+    """ Generate an AST for example.cmf """
+    with open("../grammar.ebnf") as f:
+        print("Reading ../grammar.ebnf")
+        grammar = f.read()
+        with open("../../example.cmf") as f2:
+            print("Reading ../../example.cmf")
+            cmf = f2.read()
+        ast, _ = cmfc.parse(grammar, cmf)
+        return ast
+
+
+def gen_python_code(ast):
+    """ Generate example.py from an ast """
+    print("Generating example.py")
+    code = pygen.translate(ast)
+    return code
+
+
+@pytest.fixture(scope='session')
+def codegen(tmp_path_factory):
+    """ Write example.py to testdir """
+    testdir = tmp_path_factory.getbasetemp()
+    code = gen_python_code(gen_ast())
+    print(f"Writing example.py to {testdir}")
+    with open(f"{testdir}/example.py", "w") as f:
+        f.write(code)
+    sys.path.insert(0, str(testdir))
+
+
+def assert_roundtrip(msg):
+    s = msg.serialize()
+    d, _ = msg.deserialize(s)
+    assert d == msg
+
+
+def test_NewViewElement(codegen):
+    import example
+
+    msg = example.NewViewElement()
+    msg.replica_id = 5
+    msg.digest = b"somedigest"
+    assert_roundtrip(msg)
+
+    with pytest.raises(Exception):
```

Comment:
All tests are checking for any exception but we should check whether the proper exception is thrown

Reply:
Excellent point. I just found an exception I didn't expect :) Fixing...